### PR TITLE
add example and update deepspeed/FSDP docs

### PR DESCRIPTION
**Pull request by pacman100**
### What does this PR do?
1. Add SFT example and update DeepSpeed/FSDP docs.

### CODE REVIEW

File: docs/source/accelerate/deepspeed.md [hunk 1]
```diff
@@ -6,7 +6,164 @@ rendered properly in your Markdown viewer.
...
+
+The configuration file is used to set the default options when you launch the training script.
+
+```bash
+accelerate config --config_file deepspeed_config.yaml
+```
+
+You'll be asked a few questions about your setup, and configure the following arguments. In this example, you'll use ZeRO-3 and ZeRO-Offload so make sure you pick those options.
+
+```bash
```
Comment (BenjaminBossan) [inline docs/source/accelerate/deepspeed.md:26]: The example below is not quite clear to me. Is it meant to suggest which of the options I should pick?

File: docs/source/accelerate/deepspeed.md [hunk 2]
```diff
@@ -6,7 +6,164 @@ rendered properly in your Markdown viewer.
...
 
-Both of these features are supported in 🤗 Accelerate, and you can use them with 🤗 PEFT. This guide will help you learn how to use our DeepSpeed [training script](https://github.com/huggingface/peft/blob/main/examples/conditional_generation/peft_lora_seq2seq_accelerate_ds_zero3_offload.py). You'll configure the script to train a large model for conditional generation with ZeRO-3 and ZeRO-Offload.
+Both of these features are supported in 🤗 Accelerate, and you can use them with 🤗 PEFT. 
+
+# Use PEFT and DeepSpeed with ZeRO3 for finetuning large models on multiple machines and multiple nodes
+This section of guide will help you learn how to use our DeepSpeed [training script](https://github.com/huggingface/peft/blob/main/examples/sft/train.py) for performing SFT. You'll configure the script to do SFT (supervised fine-tuning) of Llama-70B model with LoRA and ZeRO-3 on 8xH100 80GB GPUs on a single machine. You can configure it to scale to multiple machines by changing the accelerate config.
```
Comment (BenjaminBossan) [inline docs/source/accelerate/deepspeed.md:12]: - Let's spell out what "SFT" means the first time it is mentioned.
- Let's already mention here the task/dataset being used in this example.
- "8xH100 80GB GPUs on a single machine" but below SLURM is mentioned. Would we really need SLURM when using a single machine?

File: docs/source/accelerate/deepspeed.md [hunk 3]
```diff
@@ -6,7 +6,164 @@ rendered properly in your Markdown viewer.
...
+same_network: true
+tpu_env: []
+tpu_use_cluster: false
+tpu_use_sudo: false
+use_cpu: false
+```
+
+## Launch command
+
+The launch command is available at [deepspeed_config.yaml](https://github.com/huggingface/peft/blob/main/examples/sft/run_peft_deepspeed.sh) and it is also shown below:
```
Comment (BenjaminBossan) [inline docs/source/accelerate/deepspeed.md:67]:
```suggestion
The launch command is available at [deepspeed_config.sh](https://github.com/huggingface/peft/blob/main/examples/sft/run_peft_deepspeed.sh) and it is also shown below:
```

File: docs/source/accelerate/deepspeed.md [hunk 4]
```diff
@@ -6,7 +6,164 @@ rendered properly in your Markdown viewer.
...
+--hub_private_repo False \
+--hub_strategy "every_save" \
+--bf16 True \
+--packing True \
+--learning_rate 1e-4 \
+--lr_scheduler_type "cosine" \
+--weight_decay 1e-4 \
+--warmup_ratio 0.0 \
+--max_grad_norm 1.0 \
+--output_dir "mistral-sft-lora-deepspeed" \
```
Comment (BenjaminBossan) [inline docs/source/accelerate/deepspeed.md:94]: Why "mistral"? Should it be "llama2"?

File: docs/source/accelerate/deepspeed.md [hunk 5]
```diff
@@ -6,7 +6,164 @@ rendered properly in your Markdown viewer.
...
+--use_flash_attn True \
+--use_peft_lora True \
+--lora_r 8 \
+--lora_alpha 16 \
+--lora_dropout 0.1 \
+--lora_target_modules "all-linear" \
+--use_4bit_quantization False
+```
+
+Notice that we are using LoRA with  rank=8, alpha=16 and targetting all linear layers. We are passing the deepspeed config file and finetuning 70B Llama model on a subset of ultrachat dataset.
```
Comment (BenjaminBossan) [inline docs/source/accelerate/deepspeed.md:110]:
```suggestion
Notice that we are using LoRA with  rank=8, alpha=16 and targeting all linear layers. We are passing the deepspeed config file and finetuning 70B Llama model on a subset of the ultrachat dataset.
```

File: docs/source/accelerate/deepspeed.md [hunk 6]
```diff
@@ -6,7 +6,164 @@ rendered properly in your Markdown viewer.
...
+
+## The important parts
+
+Let's dive a little deeper into the script so you can see what's going on, and understand how it works.
+
+The first thing to know is that the script uses DeepSpeed for distributed training as the DeepSpeed config has been passed. The `SFTTrainer` class handles all the heavy lifting of creating PEFT model using the peft config that is passed. After that when you call `trainer.train()`, Trainer internally uses 🤗 Accelerate to prepare model, optimizer and trainer using the DeepSpeed config to create DeepSpeed engine which is then trained. in this way, PEFT with DeepSpeed finetuning works. The main code snippet is below:
```
Comment (BenjaminBossan) [inline docs/source/accelerate/deepspeed.md:116]:
```suggestion
The first thing to know is that the script uses DeepSpeed for distributed training as the DeepSpeed config has been passed. The `SFTTrainer` class handles all the heavy lifting of creating the PEFT model using the peft config that is passed. After that, when you call `trainer.train()`, `SFTTrainer` internally uses 🤗 Accelerate to prepare the model, optimizer and trainer using the DeepSpeed config to create DeepSpeed engine which is then trained. The main code snippet is below:
```

File: docs/source/accelerate/deepspeed.md [hunk 7]
```diff
@@ -6,7 +6,164 @@ rendered properly in your Markdown viewer.
...
+    dataset_kwargs={
+        "append_concat_token": data_args.append_concat_token,
+        "add_special_tokens": data_args.add_special_tokens,
+    },
+    dataset_text_field=data_args.dataset_text_field,
+    max_seq_length=data_args.max_seq_length,
+)
+trainer.accelerator.print(f"{trainer.model}")
+if model_args.use_peft_lora:
+    # handle PEFT+FSDP case
```
Comment (BenjaminBossan) [inline docs/source/accelerate/deepspeed.md:137]: This could be a bit confusing since we're dealing with DeepSpeed here. I understand that this snippet is useful to have in case we want to re-use the script for FSDP but maybe we can still remove it from the docs (and leave it in the py file) to make the docs more streamlined?

File: docs/source/accelerate/deepspeed.md [hunk 8]
```diff
@@ -6,7 +6,164 @@ rendered properly in your Markdown viewer.
...
+        fsdp_plugin.auto_wrap_policy = fsdp_auto_wrap_policy(trainer.model)
+
+# train
+checkpoint = None
+if training_args.resume_from_checkpoint is not None:
+    checkpoint = training_args.resume_from_checkpoint
+trainer.train(resume_from_checkpoint=checkpoint)
+
+# saving final model
+if trainer.is_fsdp_enabled:
```
Comment (BenjaminBossan) [inline docs/source/accelerate/deepspeed.md:152]: Same comment about FSDP.

File: docs/source/accelerate/deepspeed.md [hunk 9]
```diff
@@ -6,7 +6,164 @@ rendered properly in your Markdown viewer.
...
+    trainer.accelerator.state.fsdp_plugin.set_state_dict_type("FULL_STATE_DICT")
+trainer.save_model()
+```
+
+## Memory usage
+
+In the above example, the memory consumed per GPU is xx GB as seen in the screenshot below:
```
Comment (BenjaminBossan) [inline docs/source/accelerate/deepspeed.md:159]: Screenshot missing.

**Review by BenjaminBossan — COMMENTED** [review]
This is a great guide, thanks a lot Sourab. I already took a look at it, not sure if it's ready for review yet :)

Apart from the comments I left, do you know some common mistakes that are easy to make when using PEFT + DS? Or some PEFT features that won't work when combined with DS (like merging weights)? Maybe those could be added to the troubleshooting section and a link added in this guide.

File: docs/source/accelerate/deepspeed.md (hunk 2 above)
Comment (pacman100) [reply docs/source/accelerate/deepspeed.md:12]: Below it is mentioned in case one wants to use multiple machines

File: docs/source/accelerate/fsdp.md [hunk 10]
```diff
@@ -6,123 +6,174 @@ rendered properly in your Markdown viewer.
...
+--use_peft_lora True \
+--lora_r 8 \
+--lora_alpha 16 \
+--lora_dropout 0.1 \
+--lora_target_modules "all-linear" \
+--use_4bit_quantization False
+```
 
-The script also creates a configuration corresponding to the 🤗 PEFT method you're using. For LoRA, you'll use [`LoraConfig`] to specify the task type, and several other important parameters such as the dimension of the low-rank matrices, the matrices scaling factor, and the dropout probability of the LoRA layers. If you want to use a different 🤗 PEFT method, replace `LoraConfig` with the appropriate [class](../package_reference/tuners).
+Notice that we are using LoRA with  rank=8, alpha=16 and targetting all linear layers. We are passing the FSDP config file and finetuning 70B Llama model on a subset of ultrachat dataset.
```
Comment (BenjaminBossan) [inline docs/source/accelerate/fsdp.md]:
```suggestion
Notice that we are using LoRA with  rank=8, alpha=16 and targeting all linear layers. We are passing the FSDP config file and finetuning the 70B Llama model on a subset of the ultrachat dataset.
```

Maybe link the dataset?

File: docs/source/accelerate/fsdp.md [hunk 11]
```diff
@@ -6,123 +6,174 @@ rendered properly in your Markdown viewer.
...
-     model_name_or_path = "t5-base"
-     base_path = "temp/data/FinancialPhraseBank-v1.0"
-+    peft_config = LoraConfig(
-         task_type=TaskType.SEQ_2_SEQ_LM, inference_mode=False, r=8, lora_alpha=32, lora_dropout=0.1
-     )
-    model = AutoModelForSeq2SeqLM.from_pretrained(model_name_or_path)
-+   model = get_peft_model(model, peft_config)
+Let's dive a little deeper into the script so you can see what's going on, and understand how it works.
+
+The first thing to know is that the script uses FSDP for distributed training as the FSDP config has been passed. The `SFTTrainer` class handles all the heavy lifting of creating PEFT model using the peft config that is passed. After that when you call `trainer.train()`, Trainer internally uses 🤗 Accelerate to prepare model, optimizer and trainer using the FSDP config to create FSDP wrapped model which is then trained. in this way, PEFT with FSDP finetuning works. The main code snippet is below:
```
Comment (BenjaminBossan) [inline docs/source/accelerate/fsdp.md:111]:
```suggestion
The first thing to know is that the script uses FSDP for distributed training as the FSDP config has been passed. The `SFTTrainer` class handles all the heavy lifting of creating PEFT model using the peft config that is passed. After that when you call `trainer.train()`, Trainer internally uses 🤗 Accelerate to prepare model, optimizer and trainer using the FSDP config to create FSDP wrapped model which is then trained. The main code snippet is below:
```

seems unnecessary

File: docs/source/accelerate/deepspeed.md [hunk 12]
```diff
@@ -6,7 +6,158 @@ rendered properly in your Markdown viewer.
...
+--use_flash_attn True \
+--use_peft_lora True \
+--lora_r 8 \
+--lora_alpha 16 \
+--lora_dropout 0.1 \
+--lora_target_modules "all-linear" \
+--use_4bit_quantization False
+```
+
+Notice that we are using LoRA with  rank=8, alpha=16 and targeting all linear layers. We are passing the deepspeed config file and finetuning 70B Llama model on a subset of the ultrachat dataset.
```
Comment (BenjaminBossan) [inline docs/source/accelerate/deepspeed.md:110]: Maybe link Ultrachat?

**Review by BenjaminBossan — APPROVED** [review]
Thanks a lot for extending the PEFT docs on DeepSpeed and FSDP. Your additions are super good :fire::fire::fire:

Also very nice that you added the example scripts to along with the docs. Some comments on those:

- Is `examples/sft/__init__.py` needed? I think it can be deleted.
- Pretty cool that you also included the multi-GPU and unsloth script. Right now, they are not referenced in the docs. Maybe a short README.md could be added to the folder to explain the individual scripts, just very briefly.

I also added some small comments to the docs, but those are just nits, not blockers. Do we also want to ask Steven for a review?

**Comment by pacman100** [comment]
> Pretty cool that you also included the multi-GPU and unsloth script. Right now, they are not referenced in the docs. Maybe a short README.md could be added to the folder to explain the individual scripts, just very briefly.

Done.
